src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                                                            ImmutableDictionary<string, string>>();
+            AddEntries(UnusedValuePreference.DiscardVariable);
+            AddEntries(UnusedValuePreference.UnusedLocalVariable);
+            return builder.ToImmutable();
+
+            void AddEntries(UnusedValuePreference preference)
+            {
+                AddEntries2(preference, isUnusedLocalAssignment: true);
+                AddEntries2(preference, isUnusedLocalAssignment: false);
+            }
+
+            void AddEntries2(UnusedValuePreference preference, bool isUnusedLocalAssignment)
+            {
+                AddEntryCore(preference, isUnusedLocalAssignment, isRemovableAssignment: true);
+                AddEntryCore(preference, isUnusedLocalAssignment, isRemovableAssignment: false);
+            }
+
+            void AddEntryCore(UnusedValuePreference preference, bool isUnusedLocalAssignment, bool isRemovableAssignment)
+            {
+                var propertiesBuilder = ImmutableDictionary.CreateBuilder<string, string>();
+
+                propertiesBuilder.Add(UnusedValuePreferenceKey, preference.ToString());
+                if (isUnusedLocalAssignment)
+                {
+                    propertiesBuilder.Add(IsUnusedLocalAssignmentKey, string.Empty);
+                }
+                if (isRemovableAssignment)
+                {
+                    propertiesBuilder.Add(IsRemovableAssignmentKey, string.Empty);
+                }
+
+                builder.Add((preference, isUnusedLocalAssignment, isRemovableAssignment), propertiesBuilder.ToImmutable());
+            }
+        }
+
+        protected abstract Location GetDefinitionLocationToFade(IOperation unusedDefinition);
+        protected abstract bool SupportsDiscard(SyntaxTree tree);
+        protected abstract Option<CodeStyleOption<UnusedValuePreference>> UnusedValueExpressionStatementOption { get; }
+        protected abstract Option<CodeStyleOption<UnusedValuePreference>> UnusedValueAssignmentOption { get; }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        // Our analysis is limited to unused expressions in a code block, hence is unaffected by changes outside the code block.
+        // Hence, we can support incremental span based method body analysis.
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(
+                compilationContext => SymbolStartAnalyzer.CreateAndRegisterActions(
+                    compilationContext, GetDefinitionLocationToFade, SupportsDiscard,
+                    UnusedValueExpressionStatementOption, UnusedValueAssignmentOption));
```

Comment:
can you just pass 'this' along here.  That way we don't have so much duplicative data flowing around?  When necessary the SymbolStartAnalyzer can just retrieve the information it needs from the DiagnosticAnalyzer.